### PR TITLE
CADC -12766 PosixMapperClient implementation

### PR DESCRIPTION
**Pull request by jburke-cadc**
The getGID and getURI methods are very similar, will try to consolidate into methods each can use.
And is the augment doing the right thing?

### CODE REVIEW

File: cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java [hunk 1]
```diff
@@ -84,29 +102,212 @@ public class PosixMapperClient {
...
+                                String.format("error parsing query results, expected 2 values, found %s: %s",
+                                        tokens.length, line));
+                    }
+                    Integer gid = Integer.parseInt(tokens[0]);
+                    GroupURI groupURI = new GroupURI(URI.create(tokens[1]));
+                    posixGroups.add(new PosixGroup(gid, groupURI));
+                }
+            }
+            return posixGroups;
+        } catch (ResourceAlreadyExistsException e) {
```
Comment (pdowler) [inline cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java:308]: let this one throw up to caller

**Review by pdowler — CHANGES_REQUESTED** [review]

File: cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java [hunk 2]
```diff
@@ -84,29 +102,212 @@ public class PosixMapperClient {
...
-    public Subject augment(Subject s) {
-        throw new UnsupportedOperationException();
+    public Subject augment(Subject subject)
+        throws ResourceNotFoundException, IOException, InterruptedException {
+        if (subject == null) {
+            throw new IllegalArgumentException("subject cannot be null");
+        }
+
+        AuthMethod authMethod = AuthenticationUtil.getAuthMethodFromCredentials(subject);
+        if (authMethod == null || AuthMethod.ANON.equals(authMethod)) {
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java]: removed the anon check.

File: cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java [hunk 3]
```diff
@@ -84,29 +102,212 @@ public class PosixMapperClient {
...
+
+        URL serviceURL = regClient.getServiceURL(resourceID, Standards.POSIX_USERMAP, authMethod);
+        if (serviceURL == null) {
+            throw new ResourceNotFoundException("service not found for " + resourceID);
+        }
+
+        Set<PosixPrincipal> posixPrincipals = subject.getPrincipals(PosixPrincipal.class);
+        Set<HttpPrincipal> httpPrincipals = subject.getPrincipals(HttpPrincipal.class);
+        if (!posixPrincipals.isEmpty() && !httpPrincipals.isEmpty()) {
+            return subject; //nothing to do ?????
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java]: moved up to fail first

File: cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java (hunk 1 above)
Comment (jburke-cadc) [reply cadc-gms/src/main/java/org/opencadc/auth/PosixMapperClient.java]: updated to throw instead of catch

**Review by pdowler — APPROVED** [review]

**Review by pdowler — APPROVED** [review]

**Pull request merged**